northstar/src/runtime/island/fs.rs

```diff
@@ -0,0 +1,358 @@
+// Copyright (c) 2021 ESRLabs
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+use super::{Container, Error};
+use crate::runtime::{config::Config, island::utils::PathExt};
+use log::debug;
+use nix::{
+    libc::makedev,
+    mount::MsFlags,
+    sys::stat::{Mode, SFlag},
+    unistd,
+    unistd::{chown, Gid, Uid},
+};
+use npk::manifest::{self, MountOption, MountOptions, Resource, Tmpfs};
+use std::path::{Path, PathBuf};
+use tempfile::TempDir;
+use tokio::{fs::symlink, task};
+
+/// The minimal version of the /dev is maintained in a tmpdir. This tmpdir
+/// must be held for the lifetime of the IslandProzess
+pub(crate) type Dev = Option<TempDir>;
+
+/// Mount systemcall instruction done in init
+#[derive(Debug)]
+pub(super) struct Mount {
+    pub source: Option<PathBuf>,
+    pub target: PathBuf,
+    pub fstype: Option<&'static str>,
+    pub flags: MsFlags,
+    pub data: Option<String>,
+}
+
+impl Mount {
+    /// Execute this mount call
+    pub(super) fn mount(&self) {
+        if !self.target.exists() {
+            panic!("Missing mount point {}", self.target.display())
+        }
+
+        nix::mount::mount(
+            self.source.as_ref(),
+            &self.target,
+            self.fstype,
+            self.flags,
+            self.data.as_deref(),
+        )
+        .unwrap_or_else(|_| panic!("Failed to mount {:?}", self));
+    }
+}
+
+/// Iterate the mounts of a container and assemble a list of `mount` calls to be
+/// performed by init.
+pub(super) async fn mounts(
```

Comment:
I'd like to keep the io close the the mount call assembly since it's tightly coupled for `persist` and `dev`. The `Result` in the signature is needed anyways since `resource` can fail because of an invalid container configuration. If someone refers to a non existent resource that's not the fault of the rt.

I'll give the doc on the fn an update.